### PR TITLE
 Introduce Foxbox API sub-module. Maintain proper services in-memory cache in a separate submodule. Use EventDispatcher for Network, Settings and WebPush sub-modules. (fixes #116). r=gmarty

### LINKED ISSUE #116: We should remove services that no longer available from IndexedDB
It becomes a __huge__ problem if IndexedDB contains something that is no longer available in this case we always think that number of services has changed and we fire corresponding events and trying to re-render UI and more...

### PR DESCRIPTION
**Pull request by azasypkin**

### CODE REVIEW

File: app/js/lib/foxbox/api.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+const p = Object.freeze({
+  settings: Symbol('settings'),
+  net: Symbol('net'),
+
+  // Private methods.
+  getURL: Symbol('getURL'),
+});
+
+export default class API {
```
Comment (azasypkin) [inline app/js/lib/foxbox/api.js]: One of the ideas I'm still forming in my head is to make API work a bit like DB - it will be initialized immediately, but since all its methods are sync they will wait until user is authenticated and box is discovered....

File: app/js/lib/foxbox/network.js [hunk 2]
```diff
@@ -177,29 +179,30 @@ export default class Network {
...
+      .then(([isLocalOnline, isRemoteOnline]) => {
+        const currentState = isLocalOnline || isRemoteOnline;
+
+        this[p.local] = isLocalOnline;
+        this[p.remote] = isRemoteOnline;
+
+        if (currentState !== previousState) {
+          this.emit('online', currentState);
+        }
+      });
```
Comment (gmarty) [inline app/js/lib/foxbox/network.js]: What if one promise hangs on for a long time? It will delay the other one. I still think it's better not to use `Promise.all()` here.

File: app/js/lib/foxbox/webpush.js [hunk 3]
```diff
@@ -1,37 +1,36 @@
...
   listenForMessages: Symbol('listenForMessages'),
-  dispatchEvent: Symbol('dispatchEvent'),
 });
 
-export default class WebPush {
-  constructor(network, settings, callback) {
-    this[p.net] = network;
+export default class WebPush extends EventDispatcher {
+  constructor(api, settings) {
+    super(['push-action']);
```
Comment (gmarty) [inline app/js/lib/foxbox/webpush.js]: nit: we need to have all the event names on the same model. I like the preterite verb model like `events-changed` or`event-state-changed`.

Comment (azasypkin) [inline app/js/lib/foxbox/webpush.js]: Yep, we need to agree on the event namings for sure. Lower case, dashes to separate words and "preterite verb model" - sounds good?

Here we can name it as ```push-message-received``` or ```push-action-received```.

File: app/js/lib/foxbox/services.js [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+    if (typeA !== 'object' || valueA === null || valueB === null) {
+      return valueA !== valueB;
+    }
+
+    return !isSimilar(valueA, valueB);
+  });
+};
+
+export default class Services extends EventDispatcher {
```
Comment (gmarty) [inline app/js/lib/foxbox/services.js]: I'm not a big fan of changing the signature of the `foxbox` instance and use `foxbox.service` for everything related to the services, Splitting the logic in multiple files makes sense for the code organisation but that shouldn't impact the publicly exposed API. It seems longer and unnecessarily complicated to me. Or maybe I just need to get used to it.
What do you think?
What else is exposed from the `foxbox` instance? (e.g. `foxbox.boxes`, `foxbox.clear()`...)?

File: app/js/lib/foxbox/api.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+/**
+ * Instance of the API class is intended to abstract consumer from the API
+ * specific details (e.g. API base URL and version). It also tracks
+ * availability of the network, API host and whether correct user session is
+ * established. If any of this conditions is not met all API requests are
+ * blocked until it's possible to perform them, so consumer doesn't have to
+ * care about these additional checks.
+ */
+export default class API {
```
Comment (gmarty) [inline app/js/lib/foxbox/api.js]: :+1: 
I like this new API class. It makes the code easier to read.

File: app/js/lib/foxbox/webpush.js (hunk 3 above)
Comment (gmarty) [inline app/js/lib/foxbox/webpush.js]: Yeah that sounds good to me.

**Comment by gmarty** [comment]
I left a few comments.
There's a lot in this PR, so the commit name should be updated to account for the bigger scope. It looks good to me.

File: app/js/lib/foxbox/services.js (hunk 4 above)
Comment (azasypkin) [inline app/js/lib/foxbox/services.js]: > What do you think?

Well, for me ```foxbox.services``` looks almost the same as ```foxbox.recipes```, so I feel good with ```foxbox.services``` :) But if you're strongly against this, I'm totally fine with making it similar to what it was before.

> What else is exposed from the foxbox instance? (e.g. foxbox.boxes, foxbox.clear()...)?

Yeah, it's box management (```.boxes```, ```selectBox```), user session management (```login```, ```logout```), notification management and aggregation of box submodules (```.recipes``` and now ```.services```).

So to summarize, at this stage I'm OK with any choice, just let me know what you prefer.

File: app/js/lib/foxbox/network.js (hunk 2 above)
Comment (azasypkin) [inline app/js/lib/foxbox/network.js]: Yeah, this is a valid concern! Let me think what to do here.

File: app/js/lib/foxbox/services.js (hunk 4 above)
Comment (gmarty) [inline app/js/lib/foxbox/services.js]: Ah yeah I forgot about `foxbox.recipes`. I need to give it some thoughts, don't change anything for now. This approach is probably the best one.

File: tests/unit/lib/foxbox/services_test.js [hunk 6]
```diff
@@ -0,0 +1,195 @@
+import { waitFor } from '../../test-utils';
```
Comment (azasypkin) [inline tests/unit/lib/foxbox/services_test.js]: Here not all cases covered, just most important and obvious. Should be good enough to start with.

**Comment by azasypkin** [comment]
Should also fix https://github.com/fxbox/app/pull/109, https://github.com/fxbox/app/issues/107 and https://github.com/fxbox/app/issues/99

**Comment by azasypkin** [comment]
I think it should be ready for review, r+ @gmarty

File: app/js/lib/foxbox/foxbox.js [hunk 7]
```diff
@@ -23,55 +19,33 @@ const p = Object.freeze({
...
     super();
 
     // Private properties.
     this[p.settings] = new Settings();
     this[p.db] = new Db();
-    this[p.net] = new Network(this[p.settings], (foxboxOnline) => {
-      this._dispatchEvent('box-online', foxboxOnline);
+
+    this[p.net] = new Network(this[p.settings]);
+    this[p.net].on('online-state-changed', (online) => {
```
Comment (azasypkin) [inline app/js/lib/foxbox/foxbox.js]: Naming of events is hard :/ Tell me if you have better names for the events!

File: app/js/lib/foxbox/foxbox.js [hunk 8]
```diff
@@ -87,15 +61,10 @@ export default class Foxbox extends Service {
...
         // Start polling.
-        this[p.settings].on('pollingEnabled', () => {
-          this.togglePolling(this[p.settings].pollingEnabled);
-        });
-        this.togglePolling(this[p.settings].pollingEnabled);
-
-        this.recipes = new Recipes({
-          settings: this[p.settings],
-          net: this[p.net],
+        this[p.settings].on('polling-setting-changed', () => {
```
Comment (azasypkin) [inline app/js/lib/foxbox/foxbox.js]: I left it here as it was to reduce affected surface for this PR is already huge :) But in the future I was thinking that services itself can trigger that...

File: app/js/lib/foxbox/foxbox.js [hunk 9]
```diff
@@ -180,32 +149,29 @@ export default class Foxbox extends Service {
         // we have no record of previous registrations, we schedule
         // a retry.
         if (!this[p.boxes].length &&
-            !this[p.settings]._localOrigin &&
```
Comment (azasypkin) [inline app/js/lib/foxbox/foxbox.js]: I'm not sure why private fields were used here.

File: app/js/lib/foxbox/foxbox.js [hunk 10]
```diff
@@ -180,32 +149,29 @@ export default class Foxbox extends Service {
...
+            !this[p.settings].localOrigin &&
+            !this[p.settings].tunnelOrigin) {
           throw new Error('Registration service did not return any boxes.');
         }
 
         if (!this[p.settings].configured) {
           this.selectBox();
         }
-
-        this._dispatchEvent('box-online', true);
```
Comment (azasypkin) [inline app/js/lib/foxbox/foxbox.js]: Actually box is not online until we receive any ping response from local or remote box. It's just registration service online, so I've removed that.

File: app/js/lib/foxbox/foxbox.js [hunk 11]
```diff
@@ -180,32 +149,29 @@ export default class Foxbox extends Service {
...
       })
       .catch((error) => {
-        if (this[p.settings]._localOrigin ||
-            this[p.settings]._tunnelOrigin) {
+        if (this[p.settings].localOrigin || this[p.settings].tunnelOrigin) {
           // Default to a previously stored box registration.
-          this._dispatchEvent('box-online', true);
```
Comment (azasypkin) [inline app/js/lib/foxbox/foxbox.js]: Same here

File: app/js/lib/foxbox/network.js [hunk 12]
```diff
@@ -172,49 +175,70 @@ export default class Network {
...
 
   /**
-   * Get a URL but don't process the response.
+   * Performs HTTP 'GET' request to the specified URL. Returns 'true' if
+   * response was successful (any of 200-299 status codes) or 'false'
+   * otherwise.
    *
    * @param {string} url The URL to send the request to.
-   * @return {Promise}
+   * @return {Promise<boolean>}
```
Comment (azasypkin) [inline app/js/lib/foxbox/network.js]: Looks like ```boolean``` is what we need from this method (and logged error if any) right now. If in the future we have rich ping response we can reconsider this.

File: app/js/lib/foxbox/network.js [hunk 13]
```diff
@@ -172,49 +175,70 @@ export default class Network {
...
+   * 'online-state-changed' event.
+   *
+   * @param {boolean} previousOnlineState Previous 'online' state.
+   * @param {Symbol} localOrRemote Symbol indicating whether we process local
+   * pong or remote one.
+   * @param {boolean} isOnline Flag that indicates whether pinged server
+   * successfully responded to ping request.
+   * @private
+   */
+  [p.onPong](previousOnlineState, localOrRemote, isOnline) {
```
Comment (azasypkin) [inline app/js/lib/foxbox/network.js]: I don't really happy about this method, alternative was to create function inside ```pingBox``` on every ping. And I don't like alternative even more...

File: app/js/lib/foxbox/services.js [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+          cache.set(service.id, service);
+
+          if (isExistingService) {
+            this.emit('service-changed', service);
+          } else {
+            servicesToAddCount++;
+          }
+        });
+
+        const servicesToRemoveCount = storedServices.length +
```
Comment (azasypkin) [inline app/js/lib/foxbox/services.js]: A bit nasty calculation, but it allows us to avoid unnecessary iterations over stored services, so we only do it when we really need to remove anything.

File: app/js/lib/foxbox/settings.js [hunk 15]
```diff
@@ -29,29 +29,26 @@ const storage = localStorage ? localStorage : {
   clear: () => {},
 };
 
-export default class Settings extends Model {
+export default class Settings extends EventDispatcher {
   constructor() {
+    super(['session-changed', 'polling-setting-changed']);
```
Comment (azasypkin) [inline app/js/lib/foxbox/settings.js]: Honestly it feels like session should not belong to ```Settings```, but that's what we have now, and I'm trying to differentiate settings from non-settings with ```-setting-``` part in the event name....

File: app/js/lib/foxbox/settings.js [hunk 16]
```diff
@@ -121,13 +111,19 @@ export default class Settings extends Model {
...
+    if (this._session === session) {
+      return;
+    }
+
+    this._session = session;
+
+    if (this._session) {
       storage.setItem(`${PREFIX}session`, this._session);
+    } else {
+      storage.removeItem(`${PREFIX}session`);
```
Comment (azasypkin) [inline app/js/lib/foxbox/settings.js]: as far as I can tell, we should always remove session from localstorage if ```session``` is not non-empty string, that's why I'm changing this. Am I missing something?

File: app/js/lib/foxbox/settings.js [hunk 17]
```diff
@@ -145,9 +141,14 @@ export default class Settings extends Model {
   }
 
   set pollingEnabled(value) {
-    value = !!value;
```
Comment (azasypkin) [inline app/js/lib/foxbox/settings.js]: I haven't found any place where we pass something non-boolean here and basically we should not.

File: app/js/views/tag-item.jsx [hunk 18]
```diff
@@ -13,28 +13,7 @@ export default class TagItem extends React.Component {
   }
 
   handleSetTag(evt) {
-    const checked = evt.target.checked;
```
Comment (azasypkin) [inline app/js/views/tag-item.jsx]: It's not properly supported right now so it just introduces maintenance burden, _but_ supporting tags is the next thing I'm going to do after watchers patch.

File: tests/unit/test-utils.js [hunk 19]
```diff
@@ -0,0 +1,10 @@
+import Rx from 'rxjs';
+
+export function waitFor(fn) {
```
Comment (azasypkin) [inline tests/unit/test-utils.js]: Tell me if you prefer class with static methods instead. I found directly exported functions even more handy, but I don't have strong opinion.

File: app/js/lib/foxbox/foxbox.js [hunk 20]
```diff
@@ -23,55 +19,33 @@ const p = Object.freeze({
...
-        this._dispatchEvent('push-action', msg);
-      });
+    this[p.api] = new API(this[p.net], this[p.settings]);
 
-    // Public properties.
-    this.recipes = null;
+    this[p.webPush] = new WebPush(this[p.api], this[p.settings]);
+    this[p.webPush].on('message-received', (msg) => {
+      this._dispatchEvent('push-message-received', msg);
+    });
```
Comment (gmarty) [inline app/js/lib/foxbox/foxbox.js]: Maybe move the events listeners to the `init()` method instead of the constructor.
This may provide some optimisation where we delay the costly init after a basic view is rendered.
We could also add a `teardown` method where we remove these event listeners.

File: app/js/lib/foxbox/foxbox.js (hunk 8 above)
Comment (gmarty) [inline app/js/lib/foxbox/foxbox.js]: Sounds good for now.

File: app/js/lib/foxbox/foxbox.js (hunk 20 above)
Comment (azasypkin) [inline app/js/lib/foxbox/foxbox.js]: > Maybe move the events listeners to the init() method instead of the constructor.

Yeah, haven't thought about this, but yeah I'll move attaching event handlers to ```init```.

File: app/js/lib/foxbox/foxbox.js (hunk 9 above)
Comment (gmarty) [inline app/js/lib/foxbox/foxbox.js]: That's because the `Settings` class needed updating.

Comment (azasypkin) [inline app/js/lib/foxbox/foxbox.js]: Hmm, I thought we had public ```localOrigin/hostName``` and ```tunnelOrigin/hostName``` properties for some time already, but nevermind :)

File: app/js/lib/foxbox/services.js [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+        }
+      });
+  }
+
+  /**
+   * Creates specific service instance using raw data returned from server. If
+   * there is no specific service class provided, BaseService class is used.
+   *
+   * @param {Object} data Service raw data.
+   * @return {BaseService}
```
Comment (gmarty) [inline app/js/lib/foxbox/services.js]: It's not always an instance of `BaseService`, even if it always extends it. Not sure if we can instruct JSDoc about that. If not, `{Object}` is the most appropriate.

File: app/js/lib/foxbox/settings.js (hunk 16 above)
Comment (gmarty) [inline app/js/lib/foxbox/settings.js]: That sounds good, yeah.

File: app/js/lib/foxbox/settings.js [hunk 22]
```diff
@@ -29,29 +29,26 @@ const storage = localStorage ? localStorage : {
...
+    this._configured = storage.getItem(`${PREFIX}configured`) !== null ?
+      storage.getItem(`${PREFIX}configured`) === 'true' : false;
+    this._localOrigin = localOrigin;
+    this._tunnelOrigin = storage.getItem(`${PREFIX}tunnelOrigin`) || '';
+    this._clientId = storage.getItem(`${PREFIX}clientId`) || '';
+    this._session = storage.getItem(`${PREFIX}session`);
+    this._skipDiscovery = storage.getItem(`${PREFIX}skipDiscovery`) === 'true';
+    this._pollingEnabled = pollingEnabled;
+    this._pushEndpoint = storage.getItem(`${PREFIX}push_endpoint`) || null;
+    this._pushPubKey = storage.getItem(`${PREFIX}push_pubKey`) || null;
```
Comment (gmarty) [inline app/js/lib/foxbox/settings.js]: These all should really be private members exposed via getters and optional setters.
If not in this PR, we should create an issue to keep a track of it.

Comment (azasypkin) [inline app/js/lib/foxbox/settings.js]: Sure, I'll file issue for that.

File: app/js/views/tag-item.jsx (hunk 18 above)
Comment (gmarty) [inline app/js/views/tag-item.jsx]: Yes. IIRC you were throwing in earlier version of this PR. Maybe a `console.error` would be good to have here too.

File: tests/foxbox-simulator/http-server.js [hunk 23]
```diff
@@ -9,4 +9,8 @@ foxboxSimulator.get('/', (req, res) => {
   res.status(200).sendFile(path.join(`${__dirname}/static/login.html`));
 });
 
+foxboxSimulator.get('/ping', (req, res) => {
+  res.status(200).end();
```
Comment (gmarty) [inline tests/foxbox-simulator/http-server.js]: It's 204 as of yesterday. See https://github.com/fxbox/foxbox/pull/426/files.

File: tests/unit/lib/foxbox/services_test.js (hunk 6 above)
Comment (gmarty) [inline tests/unit/lib/foxbox/services_test.js]: Yes, we need code coverage to know where to focus on writing new unit tests in the future.

File: tests/unit/test-utils.js (hunk 19 above)
Comment (gmarty) [inline tests/unit/test-utils.js]: I don't have any opinion either. Let's see how this works for us as we add more tests.

File: app/js/lib/foxbox/services.js [hunk 24]
```diff
@@ -0,0 +1,273 @@
...
+
+  /**
+   * Retrieves the list of the services available. Relies on IndexedDB and
+   * in-memory cache.
+   *
+   * @return {Promise<Array<BaseService>>} A promise that resolves with an
+   * array of objects.
+   */
+  getAll() {
+    return this[p.getCache]().then((cache) => Array.from(cache.values()));
```
Comment (gmarty) [inline app/js/lib/foxbox/services.js]: To properly tackle #99, I'd like to have some kind of sorting here. Maybe sort by type (lights, camera, sensors...) then by name... anything that'd make listing order consistent and predictable.
What do you think?

File: app/js/lib/foxbox/foxbox.js (hunk 7 above)
Comment (gmarty) [inline app/js/lib/foxbox/foxbox.js]: The more I think about it, the less I like the `-changed` suffix. An event means a change happened, so that sounds like a redundancy to me. Also, we should take inspiration from the browser event, in this particular case, `online` seems appropriate.
Likewise, in `message-received`, a message related event is always going to be about receiving a message, so `message` looks good enough to me.

I know I suggested preterite verb, but maybe name only event names are better.
Do you have any preferences on that?

File: app/js/lib/foxbox/network.js (hunk 12 above)
Comment (gmarty) [inline app/js/lib/foxbox/network.js]: I really like this, it makes it easy to read.

File: app/js/lib/foxbox/foxbox.js (hunk 7 above)
Comment (azasypkin) [inline app/js/lib/foxbox/foxbox.js]: >> I know I suggested preterite verb, but maybe name only event names are better.

:) Yeah, I have the same feeling honestly, just followed what we agreed on previously ;) So, remove ```preterite verb`` part and we're good? I'm fine with this.

>> Likewise, in message-received, a message related event is always going to be about receiving a message, so message looks good enough to me.

Not always :) In message app we have ```message-received``` ```message-sent``` ```message-failed-to-sent``` :) But in this case just ```message``` should be ok.

File: app/js/lib/foxbox/settings.js (hunk 15 above)
Comment (gmarty) [inline app/js/lib/foxbox/settings.js]: Settings essentially started as a wrapper around local storage. I still think that it's a good thing to have all local storage elements managed in the same class. But I understand your point of view.

File: app/js/views/tag-item.jsx (hunk 18 above)
Comment (azasypkin) [inline app/js/views/tag-item.jsx]: Sure, will have something like ```console.error('Adding new tags is not supported!')```

**Comment by gmarty** [comment]
I did the review (except the tests), but because it's a big PR I want to make sure I take the time to fully understand the changes and what it means for the architecture of the lib.
I'll do that and check the tests this afternoon.

But so far so good. That's a great job!

**Comment by azasypkin** [comment]
> I did the review (except the tests), but because it's a big PR I want to make sure I take the time to fully understand the changes and what it means for the architecture of the lib.
I'll do that and check the tests this afternoon.

Sure, take your time and thanks for the review! I won't touch this PR to until you're fully done so that you have original "snapshot".

File: app/js/lib/foxbox/foxbox.js (hunk 10 above)
Comment (gmarty) [inline app/js/lib/foxbox/foxbox.js]: I think we need an event to inform the view that a new box was detected or a previous box was disconnected.
Something like `boxes` (on the same pattern than `services` events). We can then make sure that the view is always in sync with whatever the registration server returns.

File: app/js/lib/foxbox/services.js (hunk 21 above)
Comment (azasypkin) [inline app/js/lib/foxbox/services.js]: Well I really have a mixed feelings about ```{Object}``` here, but can't find anything JS specific to advocate myself - so will make it ```{Object}``` :)

File: tests/foxbox-simulator/http-server.js (hunk 23 above)
Comment (azasypkin) [inline tests/foxbox-simulator/http-server.js]: Noted, will change!

File: app/js/lib/foxbox/foxbox.js (hunk 10 above)
Comment (azasypkin) [inline app/js/lib/foxbox/foxbox.js]: >> the view is always in sync with whatever the registration server returns.

You mean, login view, right? If so, yes, let me fix that.

File: app/js/lib/foxbox/foxbox.js (hunk 7 above)
Comment (azasypkin) [inline app/js/lib/foxbox/foxbox.js]: I think we can try to not be very strict right now, and just follow "lower case + dashes to separate words + common sense" rule and see how it goes :)

File: app/js/lib/foxbox/foxbox.js (hunk 10 above)
Comment (azasypkin) [inline app/js/lib/foxbox/foxbox.js]: So, the easiest way (to not make this PR even more bigger :)) is to introduce ```discovery``` event and call it every time registration server returns a valid response (if we override ```this[p.boxes]``` I'll fire event). Later on we can improve logic to check if box list has actually changed.

File: app/js/lib/foxbox/services.js (hunk 24 above)
Comment (azasypkin) [inline app/js/lib/foxbox/services.js]: Ideally yes.In #99 my goal was to remove that annoying re-ordering that happens all the time :) So yeah, the thing that annoyed me is fixed with this PR, and we can use #99 to implement smarter sorting when we know for sure what sorting order we want.

File: app/js/lib/foxbox/settings.js (hunk 22 above)
Comment (azasypkin) [inline app/js/lib/foxbox/settings.js]: Filed https://github.com/fxbox/app/issues/130

File: tests/unit/lib/foxbox/services_test.js [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+    dbStub.getServices.returns(Promise.resolve(dbServices));
+
+    apiStub = sinon.stub({ get: () => {} });
+
+    settingsStub = sinon.stub({ pollingInterval: 2000 });
+
+    services = new Services(dbStub, apiStub, settingsStub);
+  });
+
+  it('getAll returns all currently cached services', function(done) {
```
Comment (gmarty) [inline tests/unit/lib/foxbox/services_test.js]: nit: you're using double quotes around method names in the api test file above.

File: app/js/lib/foxbox/services.js [hunk 26]
```diff
@@ -0,0 +1,273 @@
...
+
+            if (!fetchedService) {
+              this[p.db].deleteService(storedService.data);
+              cache.delete(storedService.data.id);
+            }
+          });
+        }
+
+        if (servicesToAddCount > 0 || servicesToRemoveCount > 0) {
+          this.emit('services-changed');
```
Comment (gmarty) [inline app/js/lib/foxbox/services.js]: It's not part of this patch, but I wonder if we should consolidate the way `services-changed` and `service-changed` events work. One send data and the other doesn't. Maybe they both should.

**Comment by gmarty** [comment]
OK, I left a couple of nit/question but it looks good to me. r+

File: app/js/lib/foxbox/services.js (hunk 26 above)
Comment (azasypkin) [inline app/js/lib/foxbox/services.js]: Yep, we should discuss this in the followups. But I'd rather have ```service-added``` and ```service-removed``` instead of ```services-changed```. Yes it can be triggered a plenty of times consequently, but home view should just throttle it like it would do with any other frequent event (scrolling and etc.).
